specs/capri/MetricDriverDevelopmentGuide.md

```diff
@@ -0,0 +1,606 @@
+
+
+### **Scope of Metric Driver**
+
+Metric Driver is supposed to collect data for a list of metrics  for an instance which is passed to the driver in each call. Metric driver should return an array of Metricspec structure . Each element in the array represent one metric from the metricList supplied ,
+
+Controller takes the returned metric array and send to different targets using adapters.
+
+
+### **What is Metricspec**
+
+MericSpec structure is declared in /pkg/model/metric.go, which contains name,timestamp,value and other attributes of a metric. Below is the MetricSpec structure.
+
+----------------------------
+
+**type **MetricSpec **struct **{
+
+
+
+InstanceID string **`json:"InstanceID,omitempty"`**
+
+InstanceName string **`json:"InstanceName,omitempty"`**
+
+Job string **`json:"Job,omitempty"`**
+
+/* associator - Some metric would need specific fields to relate components.
+
+  Use case could be to query volumes of a particular pool. Attaching the related_
+  components as labels would help us to form promQl query efficiently.
+  Example: node_disk_read_bytes_total{instance="121.244.95.60"}_
+  Above query will respond with all disks associated with node 121.244.95.60_
+  Since associated components vary, we will keep a map in metric struct to denote
+  the associated component type as key and component name as value
+  Example: associator[pool]=pool1 */
+
+ Associator **map**[string]string **`json:"Associator,omitempty"`**
+
+// Following fields can be used to form a unique metric name_**
+// source -\> Node/Dock_
+ Source string **`json:"Source,omitempty"`**
+// component -\> disk/logicalVolume/VG etc_**
+Component string **`json:"Component,omitempty"`**
+// name -\> metric name -\> readRequests/WriteRequests/Latency etc_**
+Name string **`json:"Name,omitempty"`**
+// unit -\> seconds/bytes/MBs etc_**
+Unit string **`json:"Unit,omitempty"`**
+// is aggregated_**
+IsAggregated bool **`json:"IsAggregated,omitempty"`**
+
+MetricValues []Metric
+
+}
+
+type Metric struct {
+
+  Timestamp int64
+
+  Value float64
+
+}
+
+----------------------------------------------------------------
+
+
+### **Who invokes Metric driver Collect?**
+
+Dock will invoke ColelctMetrics(), which is either during a metric POST API request or a scrape request to Metric exporter by Prometheus .
+
+
+### **What Methods metric driver should implement ?**
+
+Metric driver should implement all methods defined in MetricDriver interface , which is  in drivers/drivers.go
+
+-------------------
+
+type MetricDriver interface {
+
+  _//Any initialization the volume driver does while starting._
+
+Setup() error
+
+  _//Any operation the volume driver does while stopping._
+
+Unset() error
+
+  CollectMetrics(metricList []string,instanceID string) ([]model.MetricSpec, error)
+
+  ValidateMetricsSupportList(metricList []string,resourceType string) ([]string,error)
```

Comment:
Are we going to implement this method in next stage? Because I didn't see this method in the source code.

Reply:
changesd according the new interface, plz check now